Add details for Aggregator in Metrics Spec. (#1842)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

    .set_exporter(ConsoleExporter())
```

> ## Aggregator

Reviewer: Is aggregator an actual type/interface exposed from the SDK? Or it is a configuration? Or it is a conceptual thing?

Author: Per SiG meeting, I will rework the layout to focus on Aggregation (as a conceptual thing) as a configuration of View.

The term "Aggregator" will be reserve for future PR when we need to have a concrete class/type.

Author: Updated. PTAL

An `Aggregator` computes incoming Instrument [Measurements](./api.md#measurement)
into [Aggregated Metrics](./datamodel.md#opentelemetry-protocol-data-model).

The [View](./sdk.md#view) informs the SDK on creation and configuration of
Aggregators. An [Aggregator Name](./sdk.md#aggregator-name) may be used to
refer to an Aggregator and configuration.

e.g. The View specifies the string name of an Aggregator (i.e. "Histogram")
and an optional list of configuration parameter overrides.

```c#
// Use default Histogram

> meter_provider

Reviewer: nit: the code snippet here doesn't look like C# at all, I guess the intention was `python`?

Author: Fixed. PTAL

  .add_view(
    instrument_name = "X",
    aggregator = "Histogram");

// Use Histogram with custom boundaries
meter_provider
  .add_view(
    instrument_name = "X",

>     aggregator = "Histogram",

Reviewer: Instead of requiring the _name_ and _configuration parameter_ to be separate parameters, could we give the language clients flexibility so they can find their language idiomatic approach?

I think as long as we make sure the language clients provide the same **semantic**, we'll be good.

For example, I think for .NET it makes more sense to have:
```csharp
meterProvider.AddView("X" /* instrument name */, new HistogramAggregation { Boundaries = ... });
```

Here goes my reason:
1. a string "Histogram" makes it very hard to do any compile time check
2. refactor the code becomes harder
3. in general putting a string to indicate some type is weird

Author: Agreed.

to select an Aggregator and configuration parameters.

| Instrument Kind | Default Aggregator | Configuration Parameters |
| --- | --- | --- |
| [Counter](./api.md#counter) | [Sum Aggregator](./sdk.md#sum-aggregator) | SumType = Monotonic |
| [Asynchronous Counter](./api.md#asynchronous-counter) | [Sum Aggregator](./sdk.md#sum-aggregator) | SumType = Monotonic |
| [UpDownCounter](./api.md#updowncounter) | [Sum Aggregator](./sdk.md#sum-aggregator) | SumType = Non-Monotonic |
| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | [Sum Aggregator](./sdk.md#sum-aggregator) | SumType = Non-Monotonic |
| [Asynchronous Gauge](./api.md#asynchronous-gauge) | [Last Value Aggregator](./sdk.md#last-value-aggregator) | - |
| [Histogram](./api.md#histogram) | "Best" Histogram<sup>1</sup> | - |

\[1\]: The SDK will select the "best" available Histogram Aggregator. i.e. [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator).

### Last Value Aggregator

The Last Value Aggregator collects data for the
[Gauge Metric Point](./datamodel.md#gauge).

This Aggregator does not have any configuration parameters.

This Aggregator collects:

> - The last `Measurement`.

Reviewer 2: Does it collect the last `Measurement`'s exact timestamp?  I believe this is an expectation with Gauges.

Author: Yes, you are correct. Added Timestamp as Measurements just carries the Value and not a timestamp.

    .set_exporter(ConsoleExporter())
```

## Aggregator

An `Aggregator` computes incoming Instrument [Measurements](./api.md#measurement)
into [Aggregated Metrics](./datamodel.md#opentelemetry-protocol-data-model).

The [View](./sdk.md#view) informs the SDK on creation and configuration of
Aggregators. An [Aggregator Name](./sdk.md#aggregator-name) may be used to
refer to an Aggregator and configuration.

> Implementors may choose the best idiomatic practice for their language to

Reviewer: ```suggestion
Implementors MAY choose the best idiomatic practice for their language to
```

represent the semantic of an Aggregator and configuration parameters.

e.g. The View specifies an Aggregator by string name (i.e. "Histogram").

```python
# Use Histogram with custom boundaries
meter_provider
  .add_view(
    "X", 
    aggregator="Histogram", 
    aggregator_params={"Boundaries": [0, 10, 100]}
    )
```

e.g. The View specifies an Aggregator by class/type instance.

```c#
// Use Histogram with custom boundaries
meterProviderBuilder
  .AddView(
    instrumentName: "X",
    aggregator: new HistogramAggregator(
      boundaries: new double[] { 0.0, 10.0, 100.0 }
    )
  );
```

The [View](./sdk.md#view) informs the SDK to provide correct
Instrument [Measurements](./api.md#measurement) to the correct Aggregators.

An `Aggregator` is created with optional configuration parameters
(i.e. Temporality=Delta, Boundaries=[0, 10, 100]).
Default configuration parameters will be used unless overridden by
optional configuration parameters.

An `Aggregator` has a means to "update" its internal state with incoming
Instrument [Measurement](./api.md#measurement) data.

An `Aggregator` has a means to "collect"

> [Aggregated Metric](./datamodel.md#timeseries-model) data.

Reviewer 3: mention something like.. the Collect() output will be delta or cumulative, based on the aggregator setting. (where applicable)

Author: Spec refactored to avoid implementation details, which this comment seems to be. PTAL

Reviewer 4: AFAIK we're not specifying the behavior of collect, *ONLY* what metrics need to be seen in the `Export` call to exporters.

Specifically if your internal aggregator does delta metrics for sync instruments and cumulative metrics for async instruments, that's fine as long as the *SDK Output* to exporters is cumulative or delta based on the View configuration.

We aren't detailing to the level of "collect" right now.

The "collect" action should be treated as stateful and may update its
internal state.

**TBD**:
The [View](./sdk.md#view) may inform an `Aggregator` to collect

> [Examplars](./datamodel.md#exemplars).

Reviewer 3: ```suggestion
[Exemplars](./datamodel.md#exemplars).
```

| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | [Sum Aggregator](./sdk.md#sum-aggregator) | SumType = Non-Monotonic |
| [Asynchronous Gauge](./api.md#asynchronous-gauge) | [Last Value Aggregator](./sdk.md#last-value-aggregator) | - |
| [Histogram](./api.md#histogram) | "Best" Histogram<sup>1</sup> | - |

\[1\]: The SDK will select the "best" available Histogram Aggregator. i.e. [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator).

### Last Value Aggregator

The Last Value Aggregator collects data for the
[Gauge Metric Point](./datamodel.md#gauge).

This Aggregator does not have any configuration parameters.

This Aggregator collects:

- The last `Measurement`.
- The timestamp of the last `Measurement`.

### Sum Aggregator

The Sum Aggregator collects data for the [Sum Metric Point](./datamodel.md#sums).

This Aggregator honors the following configuration parameters:

| Key | Value | Default Value | Description |
| --- | --- | --- | --- |
| SumType<sup>2</sup> | Monotonic, Non-Monotonic, Other<sup>2</sup> | Depends<sup>1</sup> | See SumType<sup>2</sup> |

> | Temporality | Delta, Cumulative | Cumulative | See [Temporality](./datamodel.md#temporality). |

Reviewer 3: When we say an aggregator has a particular temporality, what exactly does it mean?

Does it mean that the output of the aggregator (Collect() output) will be delta or cumulative? 

*OR*

Does it mean that the input to the aggregator (Update() input) is a delta or cumulative value?

Or something else..

Author: The Temporality informs the Aggregator to collect data to be reported with that Temporality as an OTLP Metric Point.

Author: There are some unspecified magic... For example...

I use a Synchronous Counter instrument and report only delta values (i.e. T0=+5, T1=+5, T2=+5) and I use a Sum Aggregator (which is set to Cumulative by default). The desire output for this Aggregator is a Metric with cumulative (T0-T2) with a sum value of 15.

The Implementation of the SDK/Aggregator will need to account for Delta-to-Cumulative (and Cumulative-to-Delta) conversions somewhere.

Author: I refactored the Spec to remove implementation details. Thus, this comment is no longer relevant. PTAL.

Reviewer 4: It means that the Metric sent to `Export` of a `MetricExporter` will be delta or cumulative.

> | Temporality | Delta, Cummulative | Cummulative | See [Temporality](./datamodel.md#temporality). |

Reviewer 3: ```suggestion
| Temporality | Delta, Cumulative | Cumulative | See [Temporality](./datamodel.md#temporality). |
```

\[1\]: Depends on the Instrument Kind. See [Default Aggregator](./sdk.md#default-aggregator).

> \[2\]: **TBD**: See [PR#320](open-telemetry/opentelemetry-proto#320) for SumType definition.

Reviewer 3: ```suggestion
\[2\]: **TBD**: See [PR#320](https://github.com/open-telemetry/opentelemetry-proto/pull/320) for SumType definition.
```

This Aggregator collects:

- The arithmetic sum of `Measurement` values.

### Explicit Bucket Histogram Aggregator

The Explicit Bucket Histogram Aggregator collects data for the
[Histogram Metric Point](./datamodel.md#histogram). It supports a set of
explicit boundary values for histogram bucketing.

This Aggregator honors the following configuration parameters:

| Key | Value | Default Value | Description |
| --- | --- | --- | --- |
| Monotonic | boolean | true | if true, non-positive values are treated as errors<sup>1</sup>. |

> | Temporality | Delta, Cummulative | Cummulative | See [Temporality](./datamodel.md#temporality). |

Reviewer: ```suggestion
| Temporality | Delta, Cumulative | Cumulative | See [Temporality](./datamodel.md#temporality). |
```

```

The [View](./sdk.md#view) informs the SDK to provide correct
Instrument [Measurements](./api.md#measurement) to the correct Aggregators.

An `Aggregator` is created with optional configuration parameters
(i.e. Temporality=Delta, Boundaries=[0, 10, 100]).
Default configuration parameters will be used unless overridden by
optional configuration parameters.

An `Aggregator` has a means to "update" its internal state with incoming
Instrument [Measurement](./api.md#measurement) data.

An `Aggregator` has a means to "collect"
[Aggregated Metric](./datamodel.md#timeseries-model) data.
The "collect" action should be treated as stateful and may update its
internal state.

**TBD**:
The [View](./sdk.md#view) may inform an `Aggregator` to collect
[Examplars](./datamodel.md#exemplars).

### Aggregator Name

The [View](./sdk.md#view) can refer to the following Aggregator by name.

| Name | Aggregator | Configuration Parameters |
| --- | --- | --- |
| Default<sup>1</sup> | Depends<sup>3</sup> | Depends<sup>3</sup> |
| None<sup>2</sup> | - | - |
| Sum | [Sum Aggregator](./sdk.md#sum-aggregator) | Depends<sup>3</sup> |
| Gauge | [Last Value Aggregator](./sdk.md#last-value-aggregator) | - |

> | Histogram | [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator) | Default |

Reviewer 4: We discussed this in the SiG.  

- The name "histogram" should be reserved to allow the SDK to pick the "best" histogram implementation.  This name won't have any configuration parameters associated with it.
- The name "explicit_histogram" or "explicit_bucket_histogram" should also be allowed, and this will take the explicit bucket configuration parameter.
- Histogram instruments will default to the "histogram" aggregator, which allows SDK to choose histogram implementation.

Let me know if I captured the discussion properly.

Author: Updated Spec. PTAL

\[1\]: The name "Default" is used to select Aggregator by Instrument Kind.
See [Default Aggregators](./sdk.md#default-aggregators).

\[2\]: The name "None" is used to specify no Aggregator or
a "virtual" Aggregator which ignores all measurements.

\[3\]: Depends on the Instrument Kind. See [Default Aggregators](./sdk.md#default-aggregators).

### Default Aggregators

The SDK MUST provide default Aggregators to support the
[Metric Points](./datamodel.md#metric-points) in the
[Metrics Data Model](./datamodel.md).

The "Default" Aggregator uses the Instrument Kind (e.g. at View registration)
to select the correct Aggregator and default configuration parameters.

| Instrument Kind | Default Aggregator | Default Configuration Parameters |
| --- | --- | --- |

> | [Counter](./api.md#counter) | [Sum Aggregator](./sdk.md#sum-aggregator) | Temporality = Delta<br>SumType = Monotonic<sup>2</sup> |

Reviewer 4: Discussed in the SiG.  Until DELTA backends/OSS is more popular, we should probably default everything to Cumulative aggregation (allowing configuration to go to delta).

Author: Updated Spec. PTAL

    .set_exporter(ConsoleExporter())
```

### Aggregation

An `Aggregation`, as configured via the [View](./sdk.md#view),

> informs the SDK on the ways and means to compute

Reviewer 3: nit: since we don't specify any "how" here, "ways and means" sounds too much.

This Aggregation does not have any configuration parameters.

#### Sum Aggregation

The Sum Aggregation informs the SDK to collect data for the
[Sum Metric Point](./datamodel.md#sums).

The default values for the configuration parameters will be set based on
the Instrument Kind (e.g. at View registration OR at first seen measurement).

| Instrument Kind | Default `SumType` | Default `Temporality` |
| --- | --- | --- |
| [Counter](./api.md#counter) | Monotonic | Cumulative |
| [Asynchronous Counter](./api.md#asynchronous-counter) | Monotonic | Cumulative |
| [UpDownCounter](./api.md#updowncounter) | Non-Monotonic | Cumulative |
| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | Non-Monotonic | Cumulative |

This Aggregation honors the following configuration parameters:

| Key | Value | Default Value | Description |
| --- | --- | --- | --- |
| SumType | Monotonic, Non-Monotonic, Other | See <sup>1</sup> | See [SumType in PR](https://github.com/open-telemetry/opentelemetry-proto/pull/320). |
| Temporality | Delta, Cumulative | See <sup>1</sup> | See [Temporality](./datamodel.md#temporality). |

\[1\]: See Default values based on Instrument Kind above.

> This Aggregation informs the SDK to collect:

Reviewer 3: probably remove this line, as its already mentioned in the very beginning of this section. (line 293)

Author: It's different. This section actually specifies the arithmetic sum.

- The arithmetic sum of `Measurement` values.

#### Last Value Aggregation

> The Last Value Aggregation informs the SDK to collects data for the

Reviewer 3: ```suggestion
The Last Value Aggregation informs the SDK to collect data for the
```

    .set_exporter(ConsoleExporter())
```

## Aggregator

An `Aggregator` computes incoming Instrument [Measurements](./api.md#measurement)

> into [Pre-Aggregated Metrics](./datamodel.md#opentelemetry-protocol-data-model).

Reviewer 5: It seems strange to call these "pre-aggregated" metrics when we're obviously aggregating them here. In what sense are these "pre-aggregated"? If it's only in the sense that "they are aggregated before they are exported", I think we should just call them "aggregated" metrics here rather than have OTLP verbiage leak into this part of the spec.

The [View](./sdk.md#view) informs the SDK on creation and configuration of

> Aggregators. An [Aggregator Name](./sdk.md#aggregator-name) may be use to

Reviewer 3: ```suggestion
Aggregators. An [Aggregator Name](./sdk.md#aggregator-name) may be used to
```

refer to an Aggregator and configuration.

e.g. The View specifies the string name of an Aggregator (i.e. "Histogram")
and an optional list of configuration parameter overrides.

```c#
// Use default Histogram
meter_provider
  .add_view(
    instrument_name = "X",
    aggregator = "Histogram");

// Use Histogram with custom boundaries and monotonic values
meter_provider
  .add_view(
    instrument_name = "X",
    aggregator = "Histogram",
    aggregator_params = new KeyValuePair<string, object>[]
    {

>       new KeyValuePair<string, object>("Monotonic", true),

Reviewer 2: Since https://github.com/open-telemetry/opentelemetry-proto/pull/320 is unsettled, let's leave out this "Monotonic" option in the example. I'd call it "SumType" to match that PR, in any case.

      new KeyValuePair<string, object>("Boundaries", new double[] { 0.0, 10.0, 100.0 }),
    });
```

The [View](./sdk.md#view) informs the SDK to provide correct
Instrument [Measurements](./api.md#measurement) to the correct Aggregators.

An `Aggregator` is created with optional configuration parameters
(i.e. Monotonic=true, Boundaries=[0, 10, 100]).

> Default configuration parameters will be use unless overridden by

Reviewer 6: ```suggestion
Default configuration parameters will be used unless overridden by
```

optional configuration parameters.

> An `Aggregator` has a means to "update" with incoming Instrument

Reviewer 6: ```suggestion
An `Aggregator` has a means to "update" its internal state with incoming Instrument
```

[Measurement](./api.md#measurement) data.

An `Aggregator` has a means to "collect"
[Pre-Aggregated Metric](./datamodel.md#timeseries-model) data.
The "collect" action should be treated as stateful and may update its
internal state.

> e.g. When "collect" is acted on a Sum Aggregator, configured for delta

Reviewer 6: ```suggestion
e.g. When "collect" is called on a Sum Aggregator configured for delta
```

Reviewer 2: I think this whole paragraph can be removed. It's verging on a model implementation. An implementation is required to support cumulative export somehow. When implementing cumulative state, there is an implied memory requirement but we don't have to say exactly how it is organized. Whether the SDK uses delta aggregators and a delta-to-cumulative arrangement (as my OTel-Go prototype has), as compared with delta-or-cumulative aggregators, is an implementation decision.

Reviewer 4: +1.   We can just focus this on the configuration allowed for aggregators on `View`s and the resulting metrics.

temporality, it will return a delta metric and reset its internal
counts and time-period window. A downstream Exporter, supporting only
cumulative temporality, may choose to 1. drop the delta metric
OR 2. support [Delta to Cumulative](./datamodel.md#sums-delta-to-cumulative)
conversion.

**TBD**:
The [View](./sdk.md#view) may inform an `Aggregator` to collect
[Examplars](./datamodel.md#exemplars).

### Aggregator Name

The [View](./sdk.md#view) can refer to the following Aggregator by name.

| Name | Aggregator | Configuration Parameters |
| --- | --- | --- |
| Default<sup>1</sup> | Depends<sup>3</sup> | Depends<sup>3</sup> |
| None<sup>2</sup> | - | - |
| Sum | [Sum Aggregator](./sdk.md#sum-aggregator) | Depends<sup>3</sup> |
| Gauge | [Last Value Aggregator](./sdk.md#last-value-aggregator) | - |
| Histogram | [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator) | Default |

> \[1\]: The name "Default" is use to select Aggregator by Instrument Kind.

Reviewer 6: ```suggestion
\[1\]: The name "Default" is used to select Aggregator by Instrument Kind.
```

See [Default Aggregators](./sdk.md#default-aggregators).

> \[2\]: The name "None" is use to specify no Aggregator or

Reviewer 6: ```suggestion
\[2\]: The name "None" is used to specify no Aggregator or
```

a "virtual" Aggregator which ignores all measurements.

\[3\]: Depends on the Instrument Kind. See [Default Aggregators](./sdk.md#default-aggregators).

### Default Aggregators

The SDK MUST provide default Aggregators to support the
[Metric Points](./datamodel.md#metric-points) in the
[Metrics Data Model](./datamodel.md).

> The "Default" Aggregator use the Instrument Kind (e.g. at View registration)

Reviewer 6: ```suggestion
The "Default" Aggregator uses the Instrument Kind (e.g. at View registration)
```

to select the correct Aggregator and default configuration parameters.

| Instrument Kind | Default Aggregator | Monotonic | Temporality | Boundaries |
| --- | --- | --- | --- | --- |
| [Counter](./api.md#counter) | [Sum Aggregator](./sdk.md#sum-aggregator) | true | Delta | - |
| [UpDownCounter](./api.md#updowncounter) | [Sum Aggregator](./sdk.md#sum-aggregator) | false | Delta | - |
| [Asynchronous Counter](./api.md#asynchronous-counter) | [Sum Aggregator](./sdk.md#sum-aggregator) | true | Cumulative | - |
| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | [Sum Aggregator](./sdk.md#sum-aggregator) | false | Cumulative | - |
| [Asynchronous Gauge](./api.md#asynchronous-gauge) | [Last Value Aggregator](./sdk.md#last-value-aggregator) | - | - | - |
| [Histogram](./api.md#histogram) | [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator) | true | Delta | Default<sup>1</sup> |

\[1\]: See Default Bucket Boundaries in [Explicit Bucket Histogram Aggregator](./sdk.md#explicit-bucket-histogram-aggregator).

### Last Value Aggregator

> The Last Value Aggregator collect data for the

Reviewer 6: ```suggestion
The Last Value Aggregator collects data for the
```

[Gauge Metric Point](./datamodel.md#gauge).

> This Aggregator does not have any configuration parameter keys.

Reviewer 6: ```suggestion
This Aggregator does not have any configuration parameters.
```

> This Aggregator reports on:

Reviewer 6: ```suggestion
This Aggregator provides:
```

- The last `Measurement`.

### Sum Aggregator

> The Sum Aggregator collect data for the [Sum Metric Point](./datamodel.md#sums).

Reviewer 6: ```suggestion
The Sum Aggregator collects data for the [Sum Metric Point](./datamodel.md#sums).
```

This Aggregator honors the following configuration parameter keys:

| Key | Value | Default | Description |
| --- | --- | --- | --- |

> | Monotonic | boolean | Depends<sup>1</sup> | if true, non-positive values are treated as errors<sup>2</sup>. |

Reviewer 6: Is it possible for Cumulative values to be reported here? I'm trying to come up with an example but consider the following sequence of observations: `[-200, -150, -150, -100]`. These values are monotonically increasing, but non-positive, and would therefore lead to an error according to this definition. Is this intentional?

Reviewer 4: Yes.  The word "monontonic" is overloaded for "non-negative monotonic".  This limitation is designed to work well with prometheus usage in practice.

| Temporality | Delta, Cummulative | Depends<sup>1</sup> | See [Temporality](./datamodel.md#temporality). |

\[1\]: Depends on the Instrument Kind. See [Default Aggregators](./sdk.md#default-aggregators).

\[2\]: Language implementation may choose best strategy for handling errors. (i.e. Logged, Discard, etc...)

> This Aggregator reports on:

Reviewer 6: ```suggestion
This Aggregator provides:
```

> - The arithmetic sum of `Measurement` values.

Reviewer 6: ```suggestion
- The sum of `Measurement` values.
```

Author: There were previous comments to add "arithmetic" or "natural" to clarify this sum.

### Explicit Bucket Histogram Aggregator

> The Explicit Bucket Histogram Aggregator collect data for the

Reviewer 6: ```suggestion
The Explicit Bucket Histogram Aggregator collects data for the
```

[Histogram Metric Point](./datamodel.md#histogram). It supports a set of
explicit boundary values for histogram bucketing.

This Aggregator honors the following configuration parameter keys:

| Key | Value | Default | Description |
| --- | --- | --- | --- |

> | Monotonic | boolean | true | if true, non-positive values are treated as errors<sup>1</sup>. |

Reviewer 6: Same as above: Are all negative values forbidden? Or do we only ever receive deltas here?

Reviewer 2: This is discussed in https://github.com/open-telemetry/opentelemetry-proto/pull/320 and hasn't been actually approved yet. The way it's been discussed, there are three types of Sum: monotonic, non-monotonic, and other (which models a histogram of measurements from the three primitive instrument API events Counter, UpDownCounter, and Gauge, IMO).

Reviewer 4: I think this question goes a little deeper.  In Prometheus 0 and "counter resets" are a thing, i.e. we always assume we start counts at 0, and we assume any drop resulted from a reset.   OTel's monotonic sum is meant to mirror OpenMetrics/Prometheus Sum.

Author: Updated spec with TBD and reference to open-telemetry/opentelemetry-proto#320.

| Temporality | Delta, Cummulative | Delta | See [Temporality](./datamodel.md#temporality). |
| Boundaries | double\[\] | Default<sup>2</sup> | Array of increasing values representing explicit bucket boundary values. |

> \[1\]: Language implementation may choose best strategy for handling errors. (i.e. Logged, Discard, etc...)

Reviewer 6: ```suggestion
\[1\]: Language implementations may choose the best strategy for handling errors. (i.e. Log, Discard, etc...)
```

\[2\]: Default Bucket Boundaries: (-&infin;, 0], (0, 5.0], (5.0, 10.0], (10.0, 25.0], (25.0, 50.0], (50.0, 75.0], (75.0, 100.0], (100.0, 250.0], (250.0, 500.0], (500.0, 1000.0], (1000.0, +&infin;)

> This Aggregator reports on:

Reviewer 6: ```suggestion
This Aggregator provides:
```

    .set_exporter(ConsoleExporter())
```

### Aggregation

> An `Aggregation`, as configured via the [View](./sdk.md#view), informs

Reviewer 3: An `Aggregation`, as configured via the [View](./sdk.md#view), informs
the SDK on the ways and means to compute [Aggregated Metrics](./datamodel.md#opentelemetry-protocol-data-model) from incoming Instrument
[Measurements](./api.md#measurement).

just rewording suggestion.

Author: updated

to select an aggregation and configuration parameters.

| Instrument Kind | Selected Aggregation |
| --- | --- |
| [Counter](./api.md#counter) | [Sum Aggregation](./sdk.md#sum-aggregation) |
| [Asynchronous Counter](./api.md#asynchronous-counter) | [Sum Aggregation](./sdk.md#sum-aggregation) |
| [UpDownCounter](./api.md#updowncounter) | [Sum Aggregation](./sdk.md#sum-aggregation) |
| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | [Sum Aggregation](./sdk.md#sum-aggregation) |
| [Asynchronous Gauge](./api.md#asynchronous-gauge) | [Last Value Aggregation](./sdk.md#last-value-aggregation) |
| [Histogram](./api.md#histogram) | [Histogram Aggregation](./sdk.md#histogram-aggregation) |

This Aggregation does not have any configuration parameters.

#### Sum Aggregation

The Sum Aggregation informs the SDK to collect data for the
[Sum Metric Point](./datamodel.md#sums).

The default values for the configuration parameters will be set based on
the Instrument Kind (e.g. at View registration OR at first seen measurement).

| Instrument Kind | Default `SumType` | Default `Temporality` |
| --- | --- | --- |
| [Counter](./api.md#counter) | Monotonic | Cummulative |

> | [Asynchronous Counter](./api.md#asynchronous-counter) | Monotonic | Cummulative |

Reviewer 3: ```suggestion
| [Asynchronous Counter](./api.md#asynchronous-counter) | Monotonic | Cumulative |
```

> | [UpDownCounter](./api.md#updowncounter) | Non-Monotonic | Cummulative |

Reviewer 3: ```suggestion
| [UpDownCounter](./api.md#updowncounter) | Non-Monotonic | Cumulative |
```

> | [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | Non-Monotonic | Cummulative |

Reviewer 3: ```suggestion
| [Asynchrounous UpDownCounter](./api.md#asynchronous-updowncounter) | Non-Monotonic | Cumulative |
```

The Explicit Bucket Histogram Aggregation informs the SDK to collect data for
the [Histogram Metric Point](./datamodel.md#histogram) using a set of
explicit boundary values for histogram bucketing.

This Aggregation honors the following configuration parameters:

| Key | Value | Default Value | Description |
| --- | --- | --- | --- |
| Monotonic | boolean | true | if true, non-positive values are treated as errors<sup>1</sup>. |
| Temporality | Delta, Cumulative | Cumulative | See [Temporality](./datamodel.md#temporality). |
| Boundaries | double\[\] | [ 0, 5, 10, 25, 50, 75, 100, 250, 500, 1000 ] | Array of increasing values representing explicit bucket boundary values.<br><br>The Default Value represents the following buckets:<br>(-&infin;, 0], (0, 5.0], (5.0, 10.0], (10.0, 25.0], (25.0, 50.0], (50.0, 75.0], (75.0, 100.0], (100.0, 250.0], (250.0, 500.0], (500.0, 1000.0], (1000.0, +&infin;) |

\[1\]: Language implementations may choose the best strategy for handling errors. (i.e. Log, Discard, etc...)

This Aggregation informs the SDK to collect:

- Count of `Measurement` values falling within explicit bucket boundaries.
- Arithmetic sum of `Measurement` values in population.

> ### Aggregator

Reviewer: Given we're not going to introduce `Aggregator` in the first stable release, maybe we want to remove it for now?

Author: I'll remove.
